cody-repo-util.el

```diff
@@ -0,0 +1,182 @@
+;;; cody-repo-util.el --- Sourcegraph Cody repository utilities -*- lexical-binding: t; -*-
+
+;; Copyright (C) 2024 Sourcegraph, Inc.
+
+;; Version: 0.1
+;; Maintainer: Steve Yegge <stevey@sourcegraph.com>
+;; URL: https://github.com/sourcegraph/emacs-cody
+;; Package-Requires: ((emacs "26.3") (cl-lib "0.5"))
+
+;;; Commentary:
+;;
+;; This package provides utility functions to interact with version control
+;; systems within Sourcegraph Cody for Emacs.
+;;
+;; Functions in this package are used to retrieve repository information,
+;; handle URL replacements, and determine the VCS type for files in a project.
+
+;;; Code:
+
+(require 'cl-lib)
+(require 'vc)
+(require 'vc-git)
+(require 'url-parse)
+(require 'url-util) ; For `url-hexify-string`
+
+(defun cody--uri-for (buffer-or-file)
+  "Convert the BUFFER-OR-FILE to a URI string.
+Handles Windows drive letters and formats them appropriately.
+If BUFFER-OR-FILE is a buffer, retrieves its associated file name.
+Returns BUFFER-OR-FILE in case of errors or if conversion is not possible,
+or `buffer-name` as a last resort, to avoid sending a nil path to the agent."
+  (let* ((buffer (if (bufferp buffer-or-file) buffer-or-file (current-buffer)))
+         (file (or (and (bufferp buffer-or-file) (buffer-file-name buffer-or-file))
+                   (when (stringp buffer-or-file) buffer-or-file)
+                   (buffer-name buffer))))
+    (condition-case err
+        (let* ((expanded (if (string-match "^[a-zA-Z]:" file)
+                             file
+                           (expand-file-name file)))
+               (normalized (replace-regexp-in-string "\\\\" "/" expanded)) ; Normalize Windows paths
+               (uri (if (file-remote-p normalized)
+                        normalized
+                      (concat "file://"
+                              (if (string-match "\\([a-zA-Z]\\):" normalized)
```

Comment:
Windows uses UNC paths too, like `\\fileserver\c$\foo\bar.txt`

So those should be `file://fileserver/c$/foo/bar.txt` or something like that, not `file:////...` although it might work.

Why are you messing with the case of the drive letter? Does it matter?

Your `"\\([a-zA-Z]\\):"` regexp is going to match URLs with ports, but I guess that doesn't happen in practice?

Reply:
This was an attempt at a faithful port of the JetBrains function [ProtocolTextDocument#uriFor](https://sourcegraph.com/github.com/sourcegraph/jetbrains@main/-/blob/src/main/kotlin/com/sourcegraph/cody/agent/protocol/ProtocolTextDocument.kt?L186:9).